### PR TITLE
Add com.github.junrrein.PDFSlicer

### PR DESCRIPTION
**Pull request by junrrein**
Submitting [PDF Slicer](https://github.com/junrrein/pdfslicer) to Flathub.

### CODE REVIEW

**Comment by junrrein** [comment]
I just discovered an error. The application can't save a document to the user's home root directory, when running under Flatpak. Directories under home e.g. "/home/user/Documents" work fine. I'll have to investigate this one.

Please put this review on hold in the meanwhile.

**Comment by TingPing** [comment]
> I just discovered an error. The application can't save a document to the user's home root directory, when running under Flatpak. Directories under home e.g. "/home/user/Documents" work fine. I'll have to investigate this one.

The application should be ported to use GtkFileChooserNative.

**Comment by TingPing** [comment]
Hmm, at a glance it already uses it, so nevermind.

**Comment by TingPing** [comment]
bot, build com.github.junrrein.PDFSlicer

**Comment by junrrein** [comment]
What it seems to be happening is that, depending on file size and/or location, the sandboxed app can't write to the destination file(path) returned by GtkFileChooserNative.

The error doesn't always occur. I found out that it almost always occurs if trying to save directly to the home root directory, and almost never if saving to any subdirectory. It almost never occurs if the file size is small (in Slicer terms, a one-page document).

I have tried using `Gio::File::copy`, `Gio::File::move` and `Gio::File::replace_contents`. The second approach is the one currently used in Slicer.

I have distilled the problematic code into its own mini program, which you can get [here](https://github.com/junrrein/flatpak-saving-file). This program uses `Gio::File::replace_contents` to write a 1.7 MB text file to a location of your choosing.

At this stage, there are two possibilities:

* I don't know how to properly use GtkFileChooserNative and/or how to save files under a sandboxed Flatpak application (just using portals, no special filesystem permissions).
* There is a bug somewhere in the stack.

Any guidance is appreciated.

**Comment by junrrein** [comment]
I should add that Slicer didn't always use GtkFileChooserNative. Until v1.0.3, it used GtkFileChooserDialog, and `--filesystem=home` sandbox permissions. Under those conditions, the bug didn't occur.

**Comment by junrrein** [comment]
I should also add that saving files works as expected when running outside of a sandbox (both for Slicer and for the test application).

**Comment by TingPing** [comment]
Does the bug happen with home permissions and native dialog?

**Comment by junrrein** [comment]
Here are the results, with some debug output, of saving a file under the root home directory:

* When saving **without** home permissions and native dialog (when the error happens):

```
Temporary file created
Temporary file written
Destination: /run/user/1000/doc/6114c29e/Untitled document.pdf

(pdfslicer:3): glibmm-CRITICAL **: 20:32:59.712: 
unhandled exception (type Glib::Error) in signal handler:
domain: g-io-error-quark
code  : 1
what  : Error opening file ?/run/user/1000/doc/6114c29e/Untitled document.pdf?: No such file or directory
```

* When saving **with** home permissions and native dialog:

```
Temporary file created
Temporary file written
Destination: /home/julian/Untitled document.pdf
Copy transferred
```

The destination path in both cases is the one provided by the dialog. It seems that when using native dialog + home permissions, the sandbox is being bypassed (and so is the problem).

**Comment by nedrichards** [comment]
I absolutely can't reproduce this when building locally (I tested with a bunch of 2mb and 10mb PDFs). What version of `flatpak` do you have installed? It may be this error that was fixed in 0.11.4 https://github.com/flatpak/flatpak/issues/1307 aside from that, the app looks great and I'd be happy to merge as is.

For the saving issue, it'd be better reported in the flatpak project or xdg-desktop-portals

**Comment by junrrein** [comment]
I improved the test application and filed a bug report[1] under xdg-desktop-portal-gtk. I'll wait and see if this issue can get resolved, since I'd rather not ship software that is broken under my own environment.

[1] https://github.com/flatpak/xdg-desktop-portal-gtk/issues/131

**Comment by nedrichards** [comment]
I'd agree that you shouldn't want to ship broken software! Happy to hold until you're satisfied.

**Comment by nedrichards** [comment]
bot, build com.github.junrrein.PDFSlicer

**Comment by junrrein** [comment]
I found out how to reliable reproduce the issue. From the bug report I linked before:

1. Open the application and open any document.
2. Save the document. Note down the filename and path used. Saving should succeed.
3. Close the application.
4. Delete the generated file.
5. Do 1-2 again, using **the same** filename and path used previously. Saving should fail.

The issue keeps happening until I log out and log in again.

**Comment by nedrichards** [comment]
Fascinating. Thanks for the awesome bug report and concise reproducer. I'll try and get some attention on that.

**Comment by nedrichards** [comment]
bot, build com.github.junrrein.PDFSlicer

File: com.github.junrrein.PDFSlicer.json [hunk 1]
```diff
@@ -162,8 +162,8 @@
                 {
                     "type": "git",
                     "url": "/home/julian/Programacion/qtcreator-projects/pdfslicer",
```
Comment (bilelmoussaoui) [inline com.github.junrrein.PDFSlicer.json]: The URL should be updated

Comment (junrrein) [reply com.github.junrrein.PDFSlicer.json]: Now that's embarrassing. Updated.

**Comment by nedrichards** [comment]
bot, build com.github.junrrein.PDFSlicer

**Comment by nedrichards** [comment]
bot, build com.github.junrrein.PDFSlicer

**Comment by nedrichards** [comment]
bot, build com.github.junrrein.PDFSlicer

**Comment by nedrichards** [comment]
bot, build com.github.junrrein.PDFSlicer

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/com.github.junrrein.PDFSlicer

**Pull request closed without merging**